### PR TITLE
feat(mqtt-proxy): support domain

### LINKED ISSUE #3964: Request help: stream_proxy(mqtt_proxy) do not support dns resolv。
### Issue description

Background:
This is the requirement for developer in our company.  It is necessary to deploy the mqtt service inside the k8s cluster with container, the conventional practice in the past is to create loadbalancer service to expose nodeport, but it is troublesome for operation and maintenance. The current api entry is unified by apisix, and apisix also has mqtt_proxy plugin, so just to try it.

During the deployment process, many problems were encountered.

FIrst question:
Apply the plugin of mqtt_proxy to the route through the apisix-dashboard, but it does not actually take effect
```
./etcdctl --endpoints=10.111.9.154:2379 get /apisix/routes/348239466273440537 --prefix
/apisix/routes/348239466273440537

{"id":"348239466273440537","create_time":1617096496,"update_time":1617255551,"uris":["/*"],"name":"mqtt_test","desc":"xxxxx","methods":["GET","HEAD","POST","PUT","DELETE","OPTIONS","PATCH"],"hosts":["mqtttest.test.com"],"vars":[],"plugins":{"mqtt-proxy":{"disable":false,"protocol_level":4,"protocol_name":"MQTT","upstream":{"ip":"172.20.250.158","port":1883}}},"upstream":{"nodes":[{"host":"mqtt.test.svc.cluster.local","port":8080,"weight":1}],"timeout":{"connect":6000,"read":6000,"send":6000},"type":"roundrobin","pass_host":"pass"},"labels":{"domain":"test.com"},"status":1}
```

Second question:
Add stream_route through api, the ip field of upstream does not support domain name, ip must be used
In view of this problem, it is not convenient to use it in k8s. It is well known that the service domain name can be used for calls within the cluster.

```
./etcdctl --endpoints=10.111.9.154:2379 get /apisix/stream_routes --prefix
/apisix/stream_routes/1
{"id":"1","plugins":{"mqtt-proxy":{"upstream":{"ip":"mqtt.test.svc.cluster.local","port":1883},"protocol_name":"MQTT","protocol_level":4}}}
```

apisix error_log
```
2021/04/01 13:27:50 [warn] 2760#2760: *168 stream [lua] upstream.lua:193: set_upstream(): upstream node has been specified, cannot be set repeatedly while prereading client data, client: 10.254.6.185, server: 0.0.0.0:9100
2021/04/01 13:27:50 [error] 2760#2760: *168 stream [lua] balancer.lua:208: load_balancer(): failed to set server peer [mqtt.test.com:1883] err: no host allowed while connecting to upstream, client: 10.254.6.185, server: 0.0.0.0:9100, bytes from/to client:0/0, bytes from/to upstream:0/0
```

### Environment

* apisix version (cmd: `apisix version`): 2.4
* OS (cmd: `uname -a`): centos & k8s

### PR DESCRIPTION
**Pull request by spacewander**
Fix #3964

Signed-off-by: spacewander <spacewanderlzx@gmail.com>

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: docs/zh/latest/plugins/mqtt-proxy.md [hunk 1]
```diff
@@ -45,7 +45,8 @@ title: mqtt-proxy
 | -------------- | ------- | ------ | ------ | ------ | ------------------------------------------------------ |
 | protocol_name  | string  | 必须   |        |        | 协议名称，正常情况下应为“ MQTT”                        |
 | protocol_level | integer | 必须   |        |        | 协议级别，MQTT `3.1.*` 应为 `4` ，MQTT `5.0` 应是`5`。 |
-| upstream.ip    | string  | 必须   |        |        | 将当前请求转发到的上游的 IP 地址                       |
+| upstream.host  | string  | 必须   |        |        | 将当前请求转发到的上游的 IP 地址或域名                  |
+| upstream.ip    | string  | 必须   |        |        | 推荐使用“host”代替。将当前请求转发到的上游的 IP 地址                       |
```
Comment (Yiyiyimu) [inline docs/zh/latest/plugins/mqtt-proxy.md:49]: since we only need one of them, naming both of them 'required' might lead to misunderstanding

**Review by nic-chen — APPROVED** [review]

File: apisix/stream/plugins/mqtt-proxy.lua [hunk 2]
```diff
@@ -159,16 +165,38 @@ function _M.preread(conf, ctx)
...
+    if conf.host_is_domain == nil then
+        conf.host_is_domain = not ipmatcher.parse_ipv4(host)
+                              and not ipmatcher.parse_ipv6(host)
+    end
+
+    if conf.host_is_domain then
+        local ip, err = core.resolver.parse_domain(host)
+        if not ip then
+            core.log.error("failed to parse host ", host, ", err: ", err)
+            return 500
```
Comment (tokers) [inline apisix/stream/plugins/mqtt-proxy.lua:182]: Other places in this method just call `ngx.exit(1)`, why here we need to return `500`.

File: apisix/stream/plugins/mqtt-proxy.lua [hunk 3]
```diff
@@ -159,16 +165,38 @@ function _M.preread(conf, ctx)
...
+            {host = host, port = conf.upstream.port, weight = 1},
         }
     }
 
     local ok, err = upstream.check_schema(up_conf)
     if not ok then
-        return 500, err
+        core.log.error("failed to check schema ", core.json.delay_encode(up_conf),
+                       ", err: ", err)
+        return 500
```
Comment (tokers) [inline apisix/stream/plugins/mqtt-proxy.lua:199]: Ditto.

File: apisix/stream/plugins/mqtt-proxy.lua (hunk 2 above)
Comment (spacewander) [reply apisix/stream/plugins/mqtt-proxy.lua:182]: > Note : we can't invoke ngx.exit or core.respond.exit in rewrite phase and access phase. if need to exit, just return the status and body, the plugin engine will make the exit happen with the returned status and body.

https://github.com/apache/apisix/blob/master/docs/en/latest/plugin-develop.md

We need to fix it like what we have done with the http plugins

Comment (tokers) [reply apisix/stream/plugins/mqtt-proxy.lua:182]: OK, got it.

**Review by tokers — APPROVED** [review]

**Pull request merged**